### PR TITLE
[quant] Add quantize and dequantize operators to decomposition table

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103926

Summary:
This PR tries to decompose the operators in torch.ops.quantized_decomposed namespace to more
primitive aten operators, this would free us from maintaining the semantics of the quantize/dequantize
operators, which can be expressed more precises in terms of underlying aten operators

Note: this PR just adds them to the decomposition table, we haven't enable this by default yet

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Review by SherlockNoMad — APPROVED** [review]
lgtm

File: torch/ao/quantization/fx/_decomposed.py [hunk 1]
```diff
@@ -82,15 +117,19 @@ def quantize_per_tensor_tensor(
     """
     assert zero_point.numel() == 1, f"Expecting zero_point tensor to be one element, but received : {zero_point.numel()}"
     assert scale.numel() == 1, f"Expecting scale tensor to be one element, but received : {scale.numel()}"
-    return quantize_per_tensor(input, scale.item(), zero_point.item(), quant_min, quant_max, dtype)
-
-@impl(quantized_decomposed_lib, "quantize_per_tensor.tensor", "Meta")
-def quantize_per_tensor_tensor_meta(input, scale, zero_point, quant_min, quant_max, dtype):
```
Comment (kimishpatel) [inline torch/ao/quantization/fx/_decomposed.py:88]: why do we not need meta implementation? Will it complain if you keep it because this op is getting decomposed?

File: torch/_meta_registrations.py [hunk 2]
```diff
@@ -4192,6 +4190,13 @@ def activate_meta():
                 _meta_lib_dont_use_me_use_register_meta_for_mkldnn.impl(op_overload, fn)
             elif "mkl::" in op_overload.name():
                 _meta_lib_dont_use_me_use_register_meta_for_mkl.impl(op_overload, fn)
+            elif "quantized_decomposed::" in op_overload.name():
+                # from torch.ao.quantization.fx._decomposed import quantized_decomposed_lib  # noqa: F401
+
+                _QUANTIZED_DECOMPOSED_LIB = torch.library.Library(
+                    "quantized_decomposed", "IMPL", "Meta"
+                )
+                _QUANTIZED_DECOMPOSED_LIB.impl(op_overload, fn)
```
Comment (kimishpatel) [inline torch/_meta_registrations.py:4199]: I dont follow why this change is needed? I see you removed meta imlementations and you seem to be adding those here, but what is the "fn" here?

File: test/quantization/pt2e/test_quantize_pt2e.py [hunk 3]
```diff
@@ -1572,6 +1585,131 @@ def test_qat_conv_bn_relu_numerics(self):
...
+
+            # this does not work if we use torch.ops.quantized_decomposed.quantize_per_tensor op,
+            # I think it is something related to
+            # FakeTensorMode not being able to work with overloaded ops properly
+            # e.g. op torch.ops.quantized_decomposed.quantize_per_tensor is an overloaded
+            # op that accepts both float scale and Tensor scale, the scale argument in the
+            # graph is a Tensor, but the op we are using is torch.ops.quantized_decomposed.quantize_per_tensor
+            # and it is being traced as torch.ops.quantized_decomopsed.quantize_per_tensor.tensor
+            with FakeTensorMode(allow_non_fake_inputs=True) as mode:
+                fake_x = mode.from_tensor(torch.rand((1, 3, 224, 224)).to(memory_format=torch.channels_last))
```
Comment (kimishpatel) [inline test/quantization/pt2e/test_quantize_pt2e.py:1704]: why channels_last?

File: test/quantization/pt2e/test_quantize_pt2e.py [hunk 4]
```diff
@@ -1572,6 +1585,131 @@ def test_qat_conv_bn_relu_numerics(self):
...
+                out = m(fake_x)
+
+            def compile_fx(m: torch.fx.GraphModule, example_inputs: Any):
+                return m
+
+            run = torch._dynamo.optimize(compile_fx, nopython=False)(m)
```
[the remark this answers is not in the document]
Comment (kimishpatel) [reply test/quantization/pt2e/test_quantize_pt2e.py:1710]: If this is to retrace so as to get decomps, then retracing is not a requirement for quantization?

**Review by kimishpatel — CHANGES_REQUESTED** [review]
What kind of metadata appears on the decomposed nodes?

Other question I have is, why we need this decomp to be registered with dynamo/export? We are not re-exporting the model, right?

File: test/quantization/pt2e/test_quantize_pt2e.py [hunk 5]
```diff
@@ -1572,6 +1585,131 @@ def test_qat_conv_bn_relu_numerics(self):
...
+                ns.call_function(torch.ops.quantized_decomposed.dequantize_per_tensor.default),
+                ns.call_function(torch.ops.aten.convolution.default),
+                ns.call_function(torch.ops.quantized_decomposed.dequantize_per_tensor.default),
+            ]
+            self.checkGraphModuleNodes(
+                m,
+                expected_node_list=node_list,
+                expected_node_occurrence=node_occurrence
+            )
+            m = make_fx(m, decomposition_table=quant_decomp)(*copy.deepcopy(example_inputs))
```
Comment (kimishpatel) [inline test/quantization/pt2e/test_quantize_pt2e.py:1636]: So we should ahve the decomposition, right out of convert, hence my comment around not needing to register decomps for q/dq. This is something quant workflow needs to handle.

File: test/quantization/pt2e/test_quantize_pt2e.py (hunk 3 above)
Comment (jerryzh168) [reply test/quantization/pt2e/test_quantize_pt2e.py:1704]: this is test case from Intel, they said this is broken when we first landed this decomposition PR

File: test/quantization/pt2e/test_quantize_pt2e.py (hunk 5 above)
Comment (jerryzh168) [reply test/quantization/pt2e/test_quantize_pt2e.py:1636]: makes sense, I'll rewrite this to be a rewrite instead of a decomposition

File: torch/_meta_registrations.py (hunk 2 above)
Comment (jerryzh168) [reply torch/_meta_registrations.py:4199]: my understanding is that when people do register_decomposition, meta kernel is auto generated or something, but I'm planning to revert this since I want to do rewrite instead of decompositions

**Comment by jerryzh168** [comment]
landed a different PR: https://github.com/pytorch/pytorch/pull/104395
